examples/cfd/swe_nonlinear_pino/train_utils/losses.py

```diff
@@ -0,0 +1,330 @@
+import numpy as np
+from modulus.models.layers.spectral_layers import fourier_derivatives
+
+
+class LpLoss(object):
+    "Relative MSE loss used for data loss with multiple equations"
+
+    def __init__(self, d=2, p=2, size_average=True, reduction=True):
+        super(LpLoss, self).__init__()
+
+        # Dimension and Lp-norm type are postive
+        assert d > 0 and p > 0
+
+        self.d = d
+        self.p = p
+        self.reduction = reduction
+        self.size_average = size_average
+
+    def rel(self, x, y):
+        num_examples = x.size()[0]
+
+        diff_norms = torch.norm(
+            x.reshape(num_examples, -1) - y.reshape(num_examples, -1), self.p, 1
+        )
+        y_norms = torch.norm(y.reshape(num_examples, -1), self.p, 1)
+
+        if self.reduction:
+            if self.size_average:
+                return torch.mean(diff_norms / y_norms)
+            else:
+                return torch.sum(diff_norms / y_norms)
+
+        return diff_norms / y_norms
+
+    def __call__(self, x, y):
+        return self.rel(x, y)
+
+
+def swe_loss(s_pred, s_true, H=1.0, use_sum=True):
+    "Data Loss"
+    h_pred = s_pred[..., 0] - H
+    u_pred = s_pred[..., 1]
+    v_pred = s_pred[..., 2]
+    h_true = s_true[..., 0] - H
+    u_true = s_true[..., 1]
+    v_true = s_true[..., 2]
+    lploss = LpLoss(size_average=True)
+    loss_h = lploss(h_pred, h_true)
+    loss_u = lploss(u_pred, u_true)
+    loss_v = lploss(v_pred, v_true)
+    loss_s = torch.stack([loss_h, loss_u, loss_v], dim=-1)
+    if use_sum:
+        data_loss = torch.sum(loss_s)
+    else:
+        data_loss = torch.mean(loss_s)
+    return data_loss
+
+
+def ic_loss(out, s0):
+    "Initial Condition Loss"
+    batchsize = out.size(0)
+    nx = out.size(1)
+    ny = out.size(2)
+    nt = out.size(3)
+    s = out.reshape(batchsize, nx, ny, nt, 3)
+
+    lploss = LpLoss(size_average=True)
+    s_ic = s[..., 0, 0].reshape(s0.shape)
+    loss_ic = lploss(s_ic, s0)
+    return loss_ic
+
+
+def FDM_swe_nonlin(h, u, v, D=1, g=1.0, nu=1.0e-3, device=0):
+    "Original method in paper to calculate fourier derivatives"
+    batchsize = u.size(0)
+    nx = u.size(1)
+    ny = u.size(2)
+    nt = u.size(3)
+    h = h.reshape(batchsize, nx, ny, nt)
+    u = u.reshape(batchsize, nx, ny, nt)
+    v = v.reshape(batchsize, nx, ny, nt)
+    dt = D / (nt - 1)
+    dx = D / (nx)
+
+    # Variables to differentiate
+    hu = h * u
+    hv = h * v
+    huu = h * u**2
+    huv = h * u * v
+    hvv = h * v**2
+    hh = h**2
+
+    # Compute fourier transform
+    hu_h = torch.fft.fftn(hu, dim=[1, 2])
+    hv_h = torch.fft.fftn(hv, dim=[1, 2])
+    huu_h = torch.fft.fftn(huu, dim=[1, 2])
+    huv_h = torch.fft.fftn(huv, dim=[1, 2])
+    hvv_h = torch.fft.fftn(hvv, dim=[1, 2])
+    hh_h = torch.fft.fftn(hh, dim=[1, 2])
+    u_h = torch.fft.fftn(u, dim=[1, 2])
+    v_h = torch.fft.fftn(v, dim=[1, 2])
+
+    # Wavenumbers in y-direction
+    k_max = nx // 2
+    N = nx
+    k_x = (
```

Comment:
Can this potentially be sped up by using broadcasting? 

```
import torch
import time

k_max = 50
N = 100
device = 'cuda'  

# original approach
start_time_original = time.time()
original_op = torch.cat(
    (
        torch.arange(start=0, end=k_max, step=1, device=device),
        torch.arange(start=-k_max, end=0, step=1, device=device),
    ),
    0,
).reshape(N, 1).repeat(1, N).reshape(1, N, N, 1)
time_original = time.time() - start_time_original

# optimized approach
start_time_optimized = time.time()
k_sequence = torch.cat(
    (
        torch.arange(start=0, end=k_max, step=1, device=device),
        torch.arange(start=-k_max, end=0, step=1, device=device),
    ),
    0,
)
optimized_op = k_sequence[None, :, None, None]
optimized_op = optimized_op.expand(1, N, N, 1)
time_optimized = time.time() - start_time_optimized

assert torch.equal(optimized_op, original_op), "Not match!"

print(time_original, time_optimized)
```

This is currently outputing following for me:

```
0.5667016506195068 0.0003199577331542969
``` 

Not sure if this is considering all the edge cases though. Also it is using `.expand()` but I am wondering if `k_x` and `k_y` are not modified later, it should be fine? Potentially a clone can also be created by using `.clone()` after the expand op. Thoughts?

Reply:
yeah that's great! I've replaced it with `.expand()` and a `.clone()`and the loss isn't affected during training.

Reply:
Great to see! Thanks for incorporating and testing it